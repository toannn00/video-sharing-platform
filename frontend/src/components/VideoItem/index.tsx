import { lazy, Suspense } from "react";
import Video from "../../types/video.type";
import { Row, Col } from "antd";
import { useMediaQuery } from "react-responsive";
import { useInView } from "react-intersection-observer";

const ReactPlayer = lazy(() => import("react-player"));

export const VideoItem = ({ video }: { video: Video }) => {
  const { title, description, url, email } = video;
  const isMobile = useMediaQuery({ maxWidth: 1280 });
  const { ref, inView } = useInView({
    threshold: 0.1,
    triggerOnce: true,
  });

  const truncate = (text: string) => {
    return text.length > 200 ? text.substring(0, 200) + "..." : text;
  };

  const renderPlayer = () => {
    if (!inView) return <div style={{ height: 240, background: "#f0f0f0" }} />;

    return (
      <Suspense
        fallback={<div style={{ height: 240, background: "#f0f0f0" }} />}
      >
        <ReactPlayer
          width={isMobile ? "100%" : 426}
          height={240}
          url={url}
          light={true}
          playing={false}
        />
      </Suspense>
    );
  };

  if (isMobile) {
    return (
      <Row gutter={10} style={{ padding: 16 }} data-testid="movie-item">
        <Col span={24}>
          <div ref={ref} style={{ textAlign: "center", marginBottom: 16 }}>
            {renderPlayer()}
          </div>
        </Col>
        <Col span={24}>
          <div style={{ textAlign: "left" }}>
            <h2 style={{ margin: "0 0 8px" }}>{title}</h2>
            <p style={{ margin: "0 0 8px" }}>
              <b>Uploaded by:</b> {email}
            </p>
            <p style={{ margin: 0 }}>
              <b>Description:</b>
            </p>
            <p style={{ margin: 0 }}>{truncate(description)}</p>
          </div>
        </Col>
      </Row>
    );
  }

  return (
    <Row gutter={10} style={{ padding: 16 }} data-testid="movie-item">
      <Col span={10}>
        <div ref={ref} style={{ textAlign: "right" }}>
          {renderPlayer()}
        </div>
      </Col>
      <Col span={10} offset={3}>
        <div style={{ textAlign: "left" }}>
          <h2 style={{ margin: "0 0 8px" }}>{title}</h2>
          <p style={{ margin: "0 0 8px" }}>
            <b>Uploaded by:</b> {email}
          </p>
          <p style={{ margin: 0 }}>
            <b>Description:</b>
          </p>
          <p style={{ margin: 0 }}>{truncate(description)}</p>
        </div>
      </Col>
    </Row>
  );
};

export default VideoItem;
