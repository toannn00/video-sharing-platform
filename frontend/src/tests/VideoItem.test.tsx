import "@testing-library/jest-dom";
import ShallowRenderer from "react-test-renderer/shallow";
import { vi } from "vitest";
import { BrowserRouter } from "react-router-dom";
import VideoItem from "../components/VideoItem";

vi.mock("react-router-dom", () => ({
  useNavigate: () => vi.fn(),
  BrowserRouter: ({ children }: { children: React.ReactNode }) => children,
}));

const mockVideo = {
  title: "test",
  description: "test",
  url: "test",
  email: "test@gmail.com",
};

describe("VideoItem", () => {
  test("should render and match the snapshot", () => {
    const renderer = ShallowRenderer.createRenderer();
    renderer.render(
      <BrowserRouter>
        <VideoItem video={mockVideo} />
      </BrowserRouter>
    );
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
