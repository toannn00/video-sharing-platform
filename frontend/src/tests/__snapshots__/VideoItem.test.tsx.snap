// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VideoItem > should render and match the snapshot 1`] = `
<VideoItem
  video={
    {
      "description": "test",
      "email": "test@gmail.com",
      "title": "test",
      "url": "test",
    }
  }
/>
`;
